app/lib/analyzer/task_sources.dart

```diff
@@ -4,24 +4,118 @@
 
 import 'dart:async';
 
+import 'package:async/async.dart';
+import 'package:gcloud/db.dart';
+
+import '../frontend/models.dart';
 import '../shared/task_scheduler.dart';
 
+import 'models.dart';
+
+/// Creates a task when a version uploaded in the past hour has no analysis yet.
 class DatastoreHeadTaskSource extends PollingTaskSource {
-  DatastoreHeadTaskSource() : super(new Duration(minutes: 10));
+  final DatastoreDB _db;
+  DateTime _lastTs;
+  DatastoreHeadTaskSource(this._db) : super(const Duration(minutes: 10));
 
   @override
   Stream<Task> pollTasks() {
-    // TODO: implement pollTasks
-    return new Stream.empty();
+    final DateTime now = new DateTime.now().toUtc();
+    final DateTime oneHourAgo = now.subtract(new Duration(hours: 1));
+    DateTime minCreated;
+    if (_lastTs == null) {
+      minCreated = oneHourAgo;
+    } else if (_lastTs.isBefore(oneHourAgo)) {
+      // more than an hour passed since the last poll
+      minCreated = _lastTs;
+    } else {
+      minCreated = oneHourAgo;
+    }
+    _lastTs = oneHourAgo;
+    final Query q = _db.query(PackageVersion)..filter('created >=', minCreated);
+    return q.run().asyncMap(_mapPackageVersion).where((v) => v != null);
+  }
+
+  Future<Task> _mapPackageVersion(Model m) async {
+    final PackageVersion pv = m;
+    final List<PackageVersionAnalysis> items = await _db.lookup([
+      _db.emptyKey
+          .append(PackageAnalysis, id: pv.package)
+          .append(PackageVersionAnalysis, id: pv.version)
+    ]);
+    if (items.first == null) {
+      return null;
+    } else {
+      return new Task(pv.package, pv.version);
+    }
   }
 }
 
+/// Creates a task when the most recent analysis is older than [afterDays] days.
+///
+/// When [analysisVersion] is set, it also checks whether the current one is
+/// newer and creates a task if needed.
 class DatastoreHistoryTaskSource extends PollingTaskSource {
-  DatastoreHistoryTaskSource() : super(new Duration(hours: 1));
+  final DatastoreDB _db;
+  final int afterDays;
+  final String analysisVersion;
+  final int _batchSize = 10;
+  int _offset = 0;
+
+  DatastoreHistoryTaskSource(this._db,
+      {this.afterDays: 30, this.analysisVersion})
+      : super(const Duration(minutes: 10));
 
   @override
   Stream<Task> pollTasks() {
-    // TODO: implement pollTasks
-    return new Stream.empty();
+    final StreamCompleter<Task> resultCompleter = new StreamCompleter();
+    final Query q = _db.query(PackageVersion)
+      ..order('-created')
+      ..offset(_offset)
+      ..limit(_batchSize);
+    _offset += _batchSize;
```

Comment:
Since there can be new `PackageVersions`s the offset/limit mechanism is not really good (also because it is very costly).

Unfortunately I haven't had the time to implement `package:gcloud` support for cursors yet.

Why do we need to limit this? Because it takes too long to call generate the list of tasks?

That's precisely why I would've prefered having infinite streams and a merging function, like
```dart
  final sources = [new HeadPollerSource(), new HistoryPollerSource(), new ManualSource()];
  await for (final Task task in combineStreams(sources)) {
    ...
  }
```

This way the `combineStreams` would listen on all incoming streams and interleave the tasks, giving each task source equal priority.